src/main/scala/com/cognite/sdk/scala/common/OAuth2.scala

```diff
@@ -0,0 +1,85 @@
+package com.cognite.sdk.scala.common
+
+import java.util.concurrent.TimeUnit
+
+import cats.Monad
+import cats.syntax.all._
+import cats.effect.{Clock, ConcurrentEffect}
+import com.cognite.sdk.scala.common.internal.{CachedResource, ConcurrentCachedObject}
+import com.softwaremill.sttp._
+import com.softwaremill.sttp.circe.asJson
+import io.circe.Decoder
+import io.circe.derivation.deriveDecoder
+
+object OAuth2 {
+  final case class ClientCredentials(
+    tokenUri: Uri,
+    clientId: String,
+    clientSecret: String,
+    scopes: List[String]
+  )
+
+  class ClientCredentialsProvider[F[_]] private (
```

Comment:
Could this be `private class`? And if, should it? I see the CachedResource did this, but still confusing (to me at least).

Reply:
I think it could! What do you find confusing about it?

Reply:
Just so used to seeing it the other way, `private` seems to come first every other place we use it in this repo.

Reply:
Ah, I understand your question now. `private` after the name of a class makes its *constructor* private, but not the class itself. This is done in ConcurrentCachedResource because the creation of a CachedResource is an effect that should be wrapped in an F[_] (which is not possible with constructors). We do that here for the same reason. I don't think it's a problem to leave the class itself public, as it might be useful if we want to add some request counters for testing, for example.